feat(react): make session requireable in `useSession`

balazsorban44 opened the pull request:
## Reasoning 💡

A living session could be a requirement for specific pages (like dashboards). If it doesn’t exist, then the user should be redirected to a page that asks them to sign in again.

Sometimes, a user might log out by accident, or by deleting cookies on purpose. If that happens (e.g. on a separate tab), then `useSession({ required: true })` should detect the absence of a session cookie and always return a non-nullable Session object type.

When `required: true` is set, the default behavior will be to redirect the user to the sign-in page. This can be overridden by an `action()` callback:

```js
const session = useSession({
  required: true,
  action() {
    // ....
  }
})
if (session.status === "Loading") return "Loading or not authenticated..."

// session.data is always defined here.
```
## Checklist 🧢

- [ ] Documentation
- [x] Tests
- [ ] Ready to be merged


## Affected issues 🎟

Fixes #1210
Related: #2269 


BREAKING CHANGE:

The `useSession` hook now returns an object. Here is how to accommodate for this change:

```diff
- const [ session, loading ] = useSession()
+ const { data: session, loading } = useSession()
```

balazsorban44 committed:
feat(react): make session requireable in useSession

balazsorban44 committed:
fix(ts): mirror types for new `useSession` API

balazsorban44 committed:
test(react): fix next-auth/react tests

balazsorban44 committed:
feat(react): expose `loading` on `useSession` for easier migration

balazsorban44 committed:
Merge branch 'next' into feat/use-session-required

balazsorban44 committed:
test(ts): fix types

balazsorban44 committed:
test(react): fix useSession test

balazsorban44 committed:
test(react): test `useSession` without `SessionProvider`

balazsorban44 committed:
test(react): make long ExpectType multiline

balazsorban44 committed:
Merge branch 'next' into feat/use-session-required

balazsorban44 committed:
chore(app): update `useSession` calls in dev app

balazsorban44 committed:
docs(react): update README.md

balazsorban44 committed:
test(react): correctly name destructured property

balazsorban44 committed:
docs(react): update `useSession()` references

balazsorban44 committed:
docs: tweak landing page

balazsorban44 committed:
fix(react): correctly show loading when `required: true`

balazsorban44 committed:
docs(react): mention SessionRequired error

balazsorban44 committed:
docs(react): explain required sessions in `useSession`

balazsorban44 committed:
docs(react): query -> request

balazsorban44 commented:
**NOTE** to reviewers:
I tried updating the documentation everywhere, but there might be still something that I forgot. Therefore it should be scanned with extreme care, as `useSession` is one of our most important APIs :eyes:

kripod commented on www/docs/getting-started/client.md:
```diff
@@ -34,24 +34,51 @@ You can use the [session callback](/configuration/callbacks#session-callback) to
+  const { data: session, status } = useSession()
 
-  if (session) {
+  if (status === "authenticated) {
```
```suggestion
if (status === "authenticated") {
```

kripod reviewed (approved):
Great work, keep it up! 🙌

balazsorban44 committed:
Update www/docs/getting-started/client.md

Co-authored-by: Kristóf Poduszló <kripod@protonmail.com>

lluia commented on src/client/react.js:
```diff
@@ -43,8 +43,33 @@ const logger = proxyLogger(_logger, __NEXTAUTH.basePath)
+  }, [requiredAndNotLoading, action])
+
+  if (requiredAndNotLoading) {
+    return { data: value.data, loading: true, status: "loading" }
```
I think we should have just `status: loading` and drop boolean flags, in this case `loading`. 

`data` seems a bit generic too, I would call it `session`.

Usually, boolean flags [just end creating technical debt](https://medium.com/geekculture/why-you-shouldnt-use-booleans-in-rest-apis-2747b187876c). 📐

balazsorban44 replied:
Yeah, I was a bit hesitant, as migration would have been simpler. But I am on board to drop the `loading` flag anyway.

I think of v4 as quite a few breaking changes but in the name of creating a good base for future improvements. Refactoring the APIs to use objects in most situations (returned values or function parameters)` will make it possible to extend almost anything with new options/fields without requiring the user to change their APIs.

lluia commented on types/internals/react.d.ts:
```diff
@@ -26,4 +26,12 @@ export interface NextAuthConfig {
-export type SessionContext = React.Context<Session>
+export type SessionContextValue<R extends boolean = false> = R extends true
+  ?
+      | { data: Session; status: "authenticated"; loading: false }
+      | { data: null; status: "loading"; loading: true }
+  :
+      | { data: Session; status: "authenticated"; loading: false }
+      | { data: null; status: "unauthenticated" | "loading"; loading: boolean }
```
Let's have a call to discuss this one, I'm a bit confused about why the ternary here 🧐

balazsorban44 replied:
This is to enforce stricter types when using `required: true`

lluia commented on types/tests/react.test.ts:
```diff
@@ -11,9 +11,25 @@ const clientSession = {
+} else {
+  // $ExpectType Session
+  session.data
+}
```
Should we have a third case here?
```ts
const session = client.useSession({ required: false })
// $ExpectType Session
session.data

balazsorban44 replied:
In the case of `required: true`, there is only two valid status values. `loading` and `authenticated`.

lluia commented on www/docs/getting-started/client.md:
```diff
@@ -34,24 +34,51 @@ You can use the [session callback](/configuration/callbacks#session-callback) to
+Due to the way how Next.js handles `getServerSideProps` and `getInitialProps`, every protected page load has to make a server-side request to check if the session is valid and then generate the requested page (SSR). This increases server load, and if you are good with making the requests from the client, there is an alternative. You can use `useSession` in a way that makes sure you always have a  ...
+
+The default behavior is to redirect the user to the sign-in page, from where - after a successful login - they will be sent back to the page they started on. You can also define an `action()` callback, if you would like to do something else.
+
```
We must document here the different values of `status` for non TS users 💅🏽

balazsorban44 replied:
Agree! Good catch.

lluia commented on www/docs/getting-started/example.md:
```diff
@@ -72,11 +69,12 @@ export default function Page() {
+### Share/configure session state
 
-To allow session state to be shared between pages - which improves performance, reduces network traffic and avoids component state changes while rendering - you can use the NextAuth.js Provider in `pages/_app.js`.
+Use the `<SessionProvider>` to allows instances of `useSession()` to share the session object across components. It also takes care of keeping the session updated and synced between tabs/windows.
```
```suggestion
Use the `<SessionProvider>` to allow instances of `useSession()` to share the session data across components. `<SessionProvider>` will also take care of keeping the session updated and synced between tabs/windows.
```
I'm not sure about this though given is not possible anymore to use `useSession` without `<SessionProvider />` 🤔

balazsorban44 replied:
Yes, we should rephrase this to be explicit that users *must* use `SessionProvider` from now on.

lluia commented on www/docs/tutorials/securing-pages-and-api-routes.md:
```diff
@@ -60,15 +64,16 @@ export default function Page() {
 
 :::tip
-This example assumes you have configured `_app.js` to pass the `session` prop through so that it's immediately available on page load to `useSession`.
+You have to configure `_app.js` to pass it the `session` prop. This way `useSession` won't show a loading state, because it already has the session available.
```
```suggestion
When you supply a `session` prop in `_app.js`, `useSession` won't show a loading state, as it'll already have the session available. In this way, you can provide a more seamless user experience.
```

0ubbe reviewed (changes requested):
Wow, very neat 🎸 💯 ✨ 

I'm leaving quite a lot of suggestions 💬 , in an effort to further polish the API, but I consider it's already very nice 💚

Let's have a call to go through it together 👍🏽

balazsorban44 committed:
Merge branch 'next' into feat/use-session-required

balazsorban44 commented on src/client/react.js:
```diff
@@ -43,8 +43,33 @@ const logger = proxyLogger(_logger, __NEXTAUTH.basePath)
-export function useSession() {
-  return React.useContext(SessionContext)
+export function useSession(options = {}) {
+  const value = React.useContext(SessionContext)
```
As explained earlier today, this helps in development.

balazsorban44 commented on src/client/react.js:
```diff
@@ -43,8 +43,33 @@ const logger = proxyLogger(_logger, __NEXTAUTH.basePath)
+
+  const { required, action } = options
+
+  const requiredAndNotLoading = required && value.status === "unauthenticated"
```
I would wait with this until TS conversion

balazsorban44 committed:
Apply suggestions from code review

Co-authored-by: Lluis Agusti <hi@llu.lu>

ndom91 reviewed (approved):
LGTM :+1:

balazsorban44 commented:
@lluia had some requests which first need to be reviewed again, when this PR is updated. 😊

0ubbe committed:
feat(client): remove `loading`

Remove `loading` value from `useSession` in V4 given we already return a `status`value which scales better to identify the different states of the session request.

0ubbe committed:
refactor: pair review suggestions

• rename `action` to `onFail`
• reword documentation
• document `useSession` return
• reword `<SessionProvider />` usage to state is mandatory

0ubbe committed:
refactor(client): auth fail handler

Rename the handler name on `useSession` when authentication didn't pass to be more accurate.

balazsorban44 committed:
Merge branch 'next' into feat/use-session-required
